### PR TITLE
chore: Add GitHub actions workflow for publishing release, streamline release process

### LINKED ISSUE #1455: Release download test hard-codes the version to download
See https://github.com/dafny-lang/dafny/blob/master/.github/workflows/release-downloads.yml#L26

### LINKED ISSUE #1906: release-downloads.yml broken on windows-latest
Happened for the first time after 3.4.2 was released: https://github.com/dafny-lang/dafny/runs/5274631951?check_suite_focus=true

Timing coincides with this change which is very suspicious: https://github.com/actions/virtual-environments/issues/4856

### PR DESCRIPTION
**Pull request by robin-aws**
Applying more automation to the release process, not just to make it easier and safer but also to unblock the last step of addressing #2010: only running the deeper test suite nightly or manually, but ensuring the release process includes automatically checking that this suite has been run and passed.

This change adds a GitHub Action workflow (`publish-release.yml`) which is triggered by pushing a `vX.X.X` tag. It builds the release zip files and reference manual and then creates a draft GitHub release. 

I've also addressed the extra post-release steps in the release process:

* Having to check in the updated `DafnyRef.pdf` - I've marked the `out/DafnyRef.dfy` location as no longer updated, as users can get this from the releases instead. The PDF itself contains a disclaimer in the abstract to this effect (built using v3.5.0, replacing https://github.com/dafny-lang/dafny/pull/1905).
* Having to check in updated markdown files or `Options.txt` - The `refman.yml` workflow now fails if any automatic changes weren't already applied by make targets like `numbers` or `options`, just like the `dotnet-format` C# code formatting check.
* Having to update `release-downloads.yml` - Fixed this workflow to not need explicit configuration of the latest version for Windows. Also addressed the fact that this workflow was previously broken for Windows. :)

Fixes #1455 
Fixes #1906

Successful [dry] runs of the affected workflows:
* `refman.yml` failing when the build creates source changes: https://github.com/dafny-lang/dafny/runs/6242229347
* Successful run of `release-downloads.yml` on this branch: https://github.com/dafny-lang/dafny/actions/runs/2250542693
* Dry run of `publish-release.yml` on my fork: https://github.com/robin-aws/dafny/runs/6223753594 (and the draft release it created: https://github.com/robin-aws/dafny/releases/tag/untagged-e2e66e3de81c3af82c31)

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: .github/workflows/publish-release.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+name: Publish Release
```
Comment (keyboardDrummer) [inline .github/workflows/publish-release.yml:1]: Maybe rename this to `Publish Draft Release`

**Review by keyboardDrummer — DISMISSED** [review]

**Comment by keyboardDrummer** [comment]
Thanks for working on this!

File: docs/DafnyRef/Types.md [hunk 2]
```diff
@@ -2915,7 +2915,7 @@ In contrast to one-dimensional arrays, there is no operation to
 convert stretches of elements from a multi-dimensional array to a
 sequence.
 
-## 15.3 Converting arrays to sequences {#sec-array-to-seq}
+## 15.3.3 Converting arrays to sequences {#sec-array-to-seq}
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply docs/DafnyRef/Types.md:2918]: Crap, this is a bug in Numbers.java I'll have to fix :( Good catch.

File: RELEASE.md [hunk 3]
```diff
@@ -17,47 +17,47 @@
...
-   jq` on macOS (see <https://stedolan.github.io/jq/download/>). Note:
-   Each upload of this step can take a couple of minutes, so be patient.
-
-5. On the GitHub releases page, edit the new release to add in the
-   release notes. Also check the box to create a new discussion based on
-   the release.
-
-6. Push the "Publish" button. If preparing a pre-release, stop here, as
+   `Source/version.cs`, and replace the "Upcoming" header in `RELEASE_NOTES.md`
+   with the the public version number. Push and cut a PR, get it approved,
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply RELEASE.md:21]: No I think what I have is right. The release process is just bumping the version number and therefore changing "Upcoming" to "<this release>", and not applying any actual changes to document.

File: docs/DafnyRef/Types.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabiomadge) [reply docs/DafnyRef/Types.md:2918]: No, that’s how it’s supposed to work. It requires the trailing`.`

File: RELEASE.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabiomadge) [reply RELEASE.md:21]: But who puts in the new **Upcoming**?

File: docs/DafnyRef/Types.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (robin-aws) [reply docs/DafnyRef/Types.md:2918]: Actually it's the lack of a terminating `.` that was throwing it off, this version is stable.

[the remark this answers is not in the document]
Comment (fabiomadge) [reply docs/DafnyRef/Types.md]: Stable, but wrong.

**Review by atomb — APPROVED** [review]
This looks great. Thanks for doing it!

**Pull request merged**